src/engine/effects/engineeffectsdelay.h

```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "engine/engineobject.h"
+
+class EngineEffectsDelay : public EngineObjectConstIn {
+  public:
+    EngineEffectsDelay()
+            : m_delaySamples(0),
+              m_oldDelaySamples(0),
+              m_delayPos(0),
+              m_delayBufferSize(0){};
+
+    virtual ~EngineEffectsDelay(){};
+
+    void setDelay(unsigned int delaySamples);
+
+    void process(const CSAMPLE* pIn, CSAMPLE* pOutput, const int iBufferSize);
```

Comment:
In fact, I see the `gsl::span` for the first time. I studied some information about that. It looks like it could be cool. However, the `process()` method overrides the method from [EngineObjectConstIn,](https://github.com/mixxxdj/mixxx/blob/d209239967f2f4709d03c05261f8fef07ba47f75/src/engine/engineobject.h#L23) so for now, it doesn't fit well into the current structure. To still preserve hierarchy, I think that the new class should be created in `engineobject.h,` which will use `gsl::span` instead of `CSAMPLE*`, from which will the `EngineEffectsDelay` inherit. Another thing to the implementation, the use of span is great when using iterator. If I'm right, for optimal use of gsl::span in `EngineEffectsDelay`, there have to be two iterators for one for cycle for input and output buffer, if we will use the "two buffers" version.

Reply:
Since #4803 just got merged, you can also use `std::span` (which should be exactly equivalent to the pointer-based implementation because it doesn't perform bounds checking unlike `gsl::span`). Since this is somewhat performance-critical code, it would make sense to use implementation without bounds checking.

transforming `void process(const CSAMPLE* pIn, CSAMPLE* pOutput, const int iBufferSize)` to `void process(std::span<CSAMPLE> in, std::span<CSAMPLE> out);` and vice versa is trivial. We can provide a virtual method in the baseclass that has a default implementation and implements one in term of the other. Then mark the pointer-based method deprecated and then we can gradually change all implementations.

> If I'm right, for optimal use of gsl::span in EngineEffectsDelay, there have to be two iterators for one for cycle for input and output buffer, if we will use the "two buffers" version.

Why do you think so? The elements within spans are mutable, you can just change them in-place. So you can just create a span over the single mutable buffer.

Reply:
> Since this is somewhat performance-critical code, it would make sense to use implementation without bounds checking.

I agree, sounds good to me of using std::span.

> Why do you think so? The elements within spans are mutable, you can just change them in-place. So you can just create a span over the single mutable buffer.

Oh, right right. I just didn't realize that while reading the documentation.